File: DAA/Lab/TW_2_QuickSort/quicksort.py
# Sort a given set of elements using the Quicksort method and determine the time required to sort the elements . Repaet the experiment to different
# values of n, the number of elements in the list to be sorted and plot a graph ,number of elemnts against time taken to execute

import random
import time


max = 10000


def get_data(array):
    for i in range(max):
        ele = random.randint(0, 80000)
        array.append(ele)


def partition(array, beg, end):
    pivot_index = beg
    pivot = array[pivot_index]

    while beg < end:
        while beg<len(array) and array[beg] <= pivot:
            beg += 1

        while array[end] > pivot:
            end -= 1

        if beg < end:
            array[beg], array[end] = array[end], array[beg]

    array[end], array[pivot_index] = array[pivot_index], array[end]

    return end


def quick_sort(array, beg, end):
    if beg < end:
        partition_index = partition(array, beg, end)

        quick_sort(array, beg, partition_index-1)
        quick_sort(array, partition_index+1, end)


array = []
get_data(array)

start_time = time.time()
quick_sort(array, 0, len(array)-1)
end_time = time.time()
print(f"time taken is :: {end_time - start_time} seconds")

File: DAA/Lab/TW_2_QuickSort/quicksort.cpp
#include <iostream>
#include <time.h>
#include <stdio.h>
#include <bits/stdc++.h>
#include <stdlib.h>
using namespace std;

#define MAX 10000
void getdata(int array[])
{
    int i;
    //generating random numbers
    for (i = 0; i < MAX; i++)
    {
        array[i] = rand();
    }
}

void swap(int *a, int *b)
{
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(int array[], int beg, int end)
{
    int pivot = array[end];
    int i = beg - 1;

    for (int j = beg; j <= end - 1; j++)
    {
        if (array[j] < pivot)
        {
            i++;
            swap(&array[i], &array[j]);
        }
    }
    swap(&array[i + 1], &array[end]);
    return i + 1;
}

void quickSort(int array[], int beg, int end)
{
    if (beg < end)
    {
        int pivot_index = partition(array, beg, end);

        quickSort(array, beg, pivot_index-1);
        quickSort(array, pivot_index+1,end);

    }
}

int main(){
    int array[MAX];
    getdata(array);
    clock_t start, end;

    /* Recording the starting clock tick.*/
    start = clock();

    quickSort(array, 0,MAX-1);

    // Recording the end clock tick.
    end = clock();

    // Calculating total time taken by the program.
    double time_taken = double(end - start) / double(CLOCKS_PER_SEC);
    cout << "Time taken by program is : " << fixed
         << time_taken << setprecision(5);
    cout << " sec " << endl;
    return 0;
}
